reflekt-plugin/src/main/kotlin/io/reflekt/plugin/generation/bytecode/SmartReflektGeneratorExtension.kt

```diff
@@ -29,18 +49,74 @@ class SmartReflektGeneratorExtension(private val messageCollector: MessageCollec
         // Get instances stored in binding context after analysis part
         val instances = c.codegen.bindingContext.getInstances() ?: throw ReflektGenerationException("Found call to Reflekt instances, but no analysis data")
 
-        TODO("Filter instances according to invokeArguments and generate bytecode")
+        val resultValues = when (invokeParts.entityType) {
+            ReflektEntity.CLASSES -> {
+                val classInstances= instances.classes
+                filterInstances(classInstances.filter { it.isSubtypeOf(setOfNotNull(invokeArguments.subType?.fqName), binding) }, invokeArguments)
+                    .map { it.genAsmType(c) }
+            }
+            ReflektEntity.OBJECTS -> {
+                val objectInstances= instances.objects
+                filterInstances(objectInstances.filter { it.isSubtypeOf(setOfNotNull(invokeArguments.subType?.fqName), binding) }, invokeArguments)
+                    .map { it.genAsmType(c) }
+            }
+            ReflektEntity.FUNCTIONS -> {
+                val functionInstances = instances.functions
+                filterInstances(functionInstances.filter { it.checkSignature(invokeArguments.subType!!, binding) }, invokeArguments)
+                    .map { it.genAsmType(c, functionInstanceGenerator) }
+            }
+        }
+
+        // Return type (e.g. List or Set)
+        val returnType = resolvedCall.candidateDescriptor.returnType!!
+        // Type of each answer (e.g KClass or Function2)
+        val returnTypeArgument = returnType.arguments.first().type.asmType(c.typeMapper)
+
+        return StackValue.functionCall(returnType.asmType(c.typeMapper), null) {
+            it.pushArray(returnTypeArgument, resultValues, invokeParts.pushItemFunction)
+            invokeParts.invokeTerminalFunction(it)
+        }
+    }
+
+    private inline fun <reified T> filterInstances(instances: List<T>, invokeArguments: SubTypesToFilters): List<T> {
+        val resultInstances = ArrayList<T>()
+        for (instance in instances) {
+            var matches = true
+            for (filter in invokeArguments.filters) {
+                with(KotlinScriptRunner(classpath)) {
+                    addImports(importChecker.filterImports(invokeArguments.imports))
+                    filter.parameters.forEach {
+                        addValue(it, instance)
+                    }
+                    if (eval(filter.body) == false) {
```

Comment:
Please, add some comments about the return type of `eval` function. Because if you don't know the details of the implementation it looks strange, I mean the first thought at the sight of such a code is to replace it into `if (!eval(filter.body))`, but it is incorrect.